### ISSUE TITLE
`!!!` doesn't work in drake plan sometimes

### ISSUE THREAD

**Issue by tmastny** [ropensci/drake#724]
This works 

``` r
library(drake)

get_data <- function(param) {
  mtcars
}

drake_plan(
  data = target(
    get_data(param),
    transform = map(param = c(1, 2, 3))
  ),
  results = target(
    bind_rows(!!!data), 
    transform = combine(data)
  )
)
#> # A tibble: 4 x 2
#>   target  command                          
#>   <chr>   <expr>                           
#> 1 data_1  get_data(1)                      
#> 2 data_2  get_data(2)                      
#> 3 data_3  get_data(3)                      
#> 4 results bind_rows(data_1, data_2, data_3)
```

<sup>Created on 2019-02-08 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

```
> sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS  10.14.2

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] drake_6.2.1.9003

loaded via a namespace (and not attached):
 [1] fansi_0.4.0         assertthat_0.2.0    utf8_1.1.4          packrat_0.5.0      
 [5] digest_0.6.18       crayon_1.3.4        R6_2.3.0            backports_1.1.3    
 [9] storr_1.2.1         magrittr_1.5.0.9000 pillar_1.3.1        cli_1.0.1          
[13] rlang_0.3.1         rstudioapi_0.9.0    tools_3.5.1         igraph_1.2.2       
[17] compiler_3.5.1      pkgconfig_2.0.2     base64url_1.4       tibble_2.0.1   
```

But on another machine:

```r
library(drake)

get_data <- function(param) {
  mtcars
}

drake_plan(
  data = target(
    get_data(param),
    transform = map(param = c(1, 2, 3))
  ),
  results = target(
    bind_rows(!!!data),
    transform = combine(data)
  )
)
#> Error in enexpr(expr): object 'data_1' not found
```

```
> sessionInfo()
R version 3.5.2 (2018-12-20)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.2

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] drake_6.2.1.9003

loaded via a namespace (and not attached):
 [1] compiler_3.5.2   backports_1.1.3  R6_2.3.0         magrittr_1.5     base64url_1.4    tools_3.5.2      pillar_1.3.1     igraph_1.2.2    
 [9] rstudioapi_0.9.0 tibble_2.0.1     yaml_2.2.0       crayon_1.3.4     digest_0.6.18    storr_1.2.1      rlang_0.3.1.9000 pkgconfig_2.0.2 
```

**Comment by wlandau** [comment]
Hmm... could be that `rlang` is at 0.3.1 on the first machine and 0.3.1.9000 on the second one. But regardless, you do not need `!!!` to achieve the correct plan. `drake` already knows about the grouping variables you define in upstream transformations.

``` r
library(drake)

get_data <- function(param) {
  mtcars
}

drake_plan(
  data = target(
    get_data(param),
    transform = map(param = c(1, 2, 3))
  ),
  results = target(
    bind_rows(data),
    transform = combine(data)
  ),
  trace = TRUE # builds intuition
)
#> # A tibble: 4 x 5
#>   target  command                                 param data   results
#>   <chr>   <expr>                                  <chr> <chr>  <chr>  
#> 1 data_1  get_data(1)                             1     data_1 <NA>   
#> 2 data_2  get_data(2)                             2     data_2 <NA>   
#> 3 data_3  get_data(3)                             3     data_3 <NA>   
#> 4 results bind_rows(list(data_1, data_2, data_3)) <NA>  <NA>   results
```

Created on 2019-02-08 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)

**Comment by tmastny** [comment]
What about functions that accept `...`, but not lists? Then we must `!!!`:

``` r
library(drake)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

.data <- tibble(
  data_1 = 1,
  data_2 = 1,
  data_3 = 1,
  data_4 = 1
)

get_data <- function(param) {
  mtcars
}

plan1 <- drake_plan(
  data = target(
    get_data(param),
    transform = map(param = c(1, 2, 3))
  ),
  results = target(
    .data %>%
      select(!!!data),
    transform = combine(data)
  )
)
plan1
#> # A tibble: 4 x 2
#>   target  command                                 
#>   <chr>   <expr>                                  
#> 1 data_1  get_data(1)                             
#> 2 data_2  get_data(2)                             
#> 3 data_3  get_data(3)                             
#> 4 results .data %>% select(data_1, data_2, data_3)

make(plan1)
#> target data_1
#> target data_2
#> target data_3
#> target results

plan2 <- drake_plan(
  data = target(
    get_data(param),
    transform = map(param = c(1, 2, 3))
  ),
  results = target(
    .data %>%
      select(data),
    transform = combine(data)
  )
)
plan2
#> # A tibble: 4 x 2
#>   target  command                                       
#>   <chr>   <expr>                                        
#> 1 data_1  get_data(1)                                   
#> 2 data_2  get_data(2)                                   
#> 3 data_3  get_data(3)                                   
#> 4 results .data %>% select(list(data_1, data_2, data_3))

make(plan2)
#> target results
#> fail results
#> Error: Target `results` failed. Call `diagnose(results)` for details. Error message:
#>   `list(data_1, data_2, data_3)` must evaluate to column positions or names, not a list
```

<sup>Created on 2019-02-08 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

## Checking `rlang`

``` r
library(drake)

get_data <- function(param) {
  mtcars
}

drake_plan(
  data = target(
    get_data(param),
    transform = map(param = c(1, 2, 3))
  ),
  results = target(
    bind_rows(!!!data),
    transform = combine(data)
  )
)
#> # A tibble: 4 x 2
#>   target  command                          
#>   <chr>   <expr>                           
#> 1 data_1  get_data(1)                      
#> 2 data_2  get_data(2)                      
#> 3 data_3  get_data(3)                      
#> 4 results bind_rows(data_1, data_2, data_3)
```

<sup>Created on 2019-02-08 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>


```
> sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS  10.14.2

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] drake_6.2.1.9003

loaded via a namespace (and not attached):
 [1] fansi_0.4.0         assertthat_0.2.0    utf8_1.1.4          packrat_0.5.0      
 [5] digest_0.6.18       crayon_1.3.4        R6_2.3.0            backports_1.1.3    
 [9] storr_1.2.1         magrittr_1.5.0.9000 pillar_1.3.1        cli_1.0.1          
[13] rlang_0.3.1.9000    rstudioapi_0.9.0    tools_3.5.1         igraph_1.2.2       
[17] compiler_3.5.1      pkgconfig_2.0.2     base64url_1.4       tibble_2.0.1  
```

**Comment by wlandau** [comment]
The following works for me with both `rlang` 0.3.1 and https://github.com/r-lib/rlang/commit/7243c6d573e3edf448f25376e855cc6fd0a52e1c.

``` r
library(drake)
params <- c(1, 2, 3)
drake_plan(
  params = list_params(!!!params),
  data = target(
    get_data(param),
    transform = map(param = c(!!!params))
  )
)
#> # A tibble: 4 x 2
#>   target command             
#>   <chr>  <expr>              
#> 1 params list_params(1, 2, 3)
#> 2 data_1 get_data(1)         
#> 3 data_2 get_data(2)         
#> 4 data_3 get_data(3)
```

Created on 2019-02-08 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)

Session info

``` r
devtools::session_info()
#> ─ Session info ──────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 3.5.0 (2018-04-23)
#>  os       Red Hat Enterprise Linux    
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/Indiana/Indianapolis
#>  date     2019-02-08                  
#> 
#> ─ Packages ──────────────────────────────────────────────────────────────
#>  ! package     * version    date       lib
#>    assertthat    0.2.0      2017-04-11 [2]
#>  R backports     1.1.2      <NA>       [2]
#>  R base64url     1.3        <NA>       [2]
#>    callr         2.0.3      2018-04-11 [2]
#>    cli           1.0.1      2018-09-25 [1]
#>    crayon        1.3.4      2017-09-16 [2]
#>    debugme       1.1.0      2017-10-22 [2]
#>    desc          1.2.0      2018-05-01 [2]
#>    devtools      2.0.1      2018-10-26 [1]
#>  R digest        0.6.15     <NA>       [2]
#>    drake       * 6.2.1.9003 2019-02-08 [1]
#>    evaluate      0.10.1     2017-06-24 [2]
#>    fansi         0.4.0      2018-10-05 [1]
#>    fs            1.2.2      2018-03-21 [2]
#>  R glue          1.2.0      <NA>       [2]
#>    htmltools     0.3.6      2017-04-28 [2]
#>  R igraph        1.2.1      <NA>       [2]
#>    knitr         1.20       2018-02-20 [2]
#>    magrittr      1.5        2014-11-22 [2]
#>    memoise       1.1.0      2017-04-21 [2]
#>    pillar        1.3.1      2018-12-15 [1]
#>    pkgbuild      1.0.2      2018-10-16 [1]
#>    pkgconfig     2.0.2      2018-08-16 [1]
#>    pkgload       1.0.2      2018-10-29 [1]
#>    prettyunits   1.0.2      2015-07-13 [2]
#>    R6            2.2.2      2017-06-17 [2]
#>  R Rcpp          0.12.16    <NA>       [2]
#>    remotes       2.0.2      2018-10-30 [1]
#>    rlang         0.3.1.9000 2019-02-08 [1]
#>    rmarkdown     1.10       2018-06-11 [2]
#>    rprojroot     1.3-2      2018-01-03 [2]
#>    sessioninfo   1.1.1      2018-11-05 [1]
#>  R storr         1.1.3      <NA>       [2]
#>    stringi       1.2.3      2018-06-12 [2]
#>    stringr       1.3.1      2018-05-10 [2]
#>    testthat      2.0.1      2018-10-13 [1]
#>    tibble        2.0.1      2019-01-12 [1]
#>    usethis       1.4.0      2018-08-14 [1]
#>    utf8          1.1.3      2018-01-03 [2]
#>    withr         2.1.2      2018-03-15 [2]
#>    yaml          2.1.19     2018-05-01 [2]
#> ...
```

**Comment by tmastny** [comment]
Thanks for checking. Something must be configured wrong on the other computer.

**Comment by wlandau** [comment]
Please let me know what you find.

From https://github.com/ropensci/drake/issues/724#issuecomment-461920676, I think I might understand why you want `!!!`. In the plan below, do you want `select(data_1, data_2, data_3)` instead of `select(list(data_1, data_2, data_3))`? If so, I recommend writing your own function to get the symbols in `list()` and ignore the targets' values.

```r
my_select <- function(.data, list) {
  # Get the names of the targets contained in data_list.
  names <- all.vars(match.call()$list)
  .data[, names]
}
```

Related: #693.

**Comment by wlandau** [comment]
Hmm... I answered in https://github.com/ropensci/drake/issues/724#issuecomment-461925613 rather hastily. Sorry about that. Your `plan1` in https://github.com/ropensci/drake/issues/724#issuecomment-461920676 is a much more elegant way to handle this.

On closer inspection, I think this functionality appeared to work before but broke when I solved #715. And it is breaking for the right reasons: symbol substitution happens before tidy evaluation. What is happening is that `bind_rows(!!!data)` becomes `bind_rows(!!!list(data_1, data_2, data_3))` and then `drake` tries to run `rlang::expr(bind_rows(!!!list(data_1, data_2, data_3)))`.

Maybe `drake` should try to imitate `!!!` instead of just inserting `list(...)` into commands. We cannot actually use `!!!` because it would conflict with existing tidy evaluation, though.

**Comment by wlandau** [comment]
Recommendation: `do.call()`. I will open a separate issue to address `your_function(data_1, data_2, data_3)` vs `your_function(list(data_1, data_2, data_3))`.

``` r
library(drake)
library(dplyr)

get_data <- function(param) {
  mtcars
}

plan <- drake_plan(
  data = target(
    get_data(param),
    transform = map(param = c(1, 2, 3))
  ),
  results = target(
    do.call(bind_rows, data), 
    transform = combine(data)
  )
)

make(plan)
#> target data_1
#> target data_2
#> target data_3
#> target results
```

<sup>Created on 2019-02-08 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by wlandau** [comment]
I am not sure #726 is going to work out, and `do.call()` does not easily fit your `plan1` that uses `select()`. Maybe we can enhance the syntax of `combine()` to let the user pick an aggregation function for each symbol. The following example is just a sketch, it is not implemented yet.

```r
plan1 <- drake_plan(
  data = target(
    get_data(param),
    transform = map(param = c(1, 2, 3))
  ),
  results = target(
    .data %>%
      data,
    transform = combine(data = select())
  )
)
# A tibble: 4 x 2
  target  command       
  <chr>   <expr>        
1 data_1  get_data(1)   
2 data_2  get_data(2)   
3 data_3  get_data(3)   
4 results .data %>% select(data_1, data_2, data_3)
```

With other arguments:

```r
plan1 <- drake_plan(
  data = target(
    get_data(param),
    transform = map(param = c(1, 2, 3))
  ),
  results = target(
    data
    transform = combine(data = min(na.rm = TRUE))
  )
)
# A tibble: 4 x 2
  target  command       
  <chr>   <expr>        
1 data_1  get_data(1)   
2 data_2  get_data(2)   
3 data_3  get_data(3)   
4 results .data %>% min(data_1, data_2, data_3, na.rm = TRUE)
```

### Advantages

- Ease of implementation
- Speed at runtime
- Compatibility with the rest of the DSL so far.

### Disadvantage

- Because of how `substitute()` works, we can only have one combining function per symbol. We cannot assign `select()` and `min()` to different instances of `data` in the same command.

**Comment by wlandau** [comment]
Well, maybe we should require the user to always supply explicit functions to `combine()` (e.g. `transform = combine(data = list())`). It breaks back compatibility with the initial DSL, but it is more internally consistent and predictable.

**Comment by wlandau** [comment]
I think the underlying need is now addressed in #729.
